designs/sam_build_cmd.rst

```diff
@@ -0,0 +1,442 @@
+| Resolve     | No Op                                                                                  |
++-------------+----------------------------------------------------------------------------------------+
+| Compile     | ``dotnet lambda package --configuration release --output-package $BUILD/package.zip``  |
++-------------+----------------------------------------------------------------------------------------+
+| Copy Source | No Op                                                                                  |
++-------------+----------------------------------------------------------------------------------------+
+
+
+Python using PIP
+^^^^^^^^^^^^^^^^
+
++-------------+---------------------------------------------------------------------------------------+
+| Action      | Command                                                                               |
++=============+=======================================================================================+
+| Resolve     | ``pip install --isolated --disable-pip-version-check -r requirements.txt -t $BUILD``  |
++-------------+---------------------------------------------------------------------------------------+
+| Compile     | No Op                                                                                 |
++-------------+---------------------------------------------------------------------------------------+
+| Copy Source | Copy all files                                                                        |
++-------------+---------------------------------------------------------------------------------------+
+
+
+Custom Build Commands
+~~~~~~~~~~~~~~~~~~~~~
+
+Users can provide custom build command for each runtime via ``.samrc`` that will be run instead of the default build
+action provided by SAM CLI. SAM CLI will invoke this command for every resource that uses this runtime. This can be any
+arbitrary program that is accessible (via ``PATH`` resolution). The program will be invoked by setting three
+environment variables:
+
+#. ``__SOURCE_DIR`` - Folder where the source files are located
+#. ``__BUILD_DIR`` - Folder where the built artifacts should be written to
+#. ``__RUNTIME`` - AWS Lambda function runtime
+#. ``__TEMPLATE`` - Path to SAM Template
+#. Current Working Directory - Same directory where ``sam build`` command was invoked from
+
+This command is expected to return an exit code of ``zero`` to indicate success. Non-zero exit codes indicate failure.
+
+Building Native Binaries
+~~~~~~~~~~~~~~~~~~~~~~~~
+To build native binaries, we need to run on an architecture and operating system that is similar to AWS Lambda. We use
+the Docker containers provided by `Docker Lambda <https://github.com/lambci/docker-lambda>`_ project to run the same
+set of commands described above on this container. We will mount source code folder and build folder into the container
```

Comment:
Do we need to install extra things for each runtime to make this work on Docker Lambda (donet cli or go cli for example)? If so, how can we do this in a way that doesn't require downloading of the same things per build?

Reply:
No, `docker-lambda` publishes a build container that has these preinstalled